File: example/xx_three/two/geom_lines.js
// geom_lines.js

var segments = 10000;

var viewer = two.ui.scene_viewer(app.ui.begin());
//two.ui.orbit_controller(viewer);

var scene = viewer.scene;

if(typeof this.state == 'undefined') {
    this.state = 1;

    var camera = viewer.camera;
    camera.fov = 27.0; camera.near = 1.0; camera.far = 4000.0;
    camera.eye.z = 2750.0;

    //this.program = app.gfx.programs.fetch('line');
    var program = app.gfx.programs.fetch('solid');

    var material = app.gfx.materials.create('lines'); var m = material;
        m.program = program;
        m.base.shader_color = two.ShaderColor.Vertex;
    

    var geometry = new two.MeshPacker();
    geometry.primitive = two.PrimitiveType.LineStrip;

    var r = 800.0; var r2 = r / 2.0;

    for(var i = 0; i < segments; i++)
    {
        var p = new two.vec3(Math.random() * r - r2, Math.random() * r - r2, Math.random() * r - r2);
        var c = new two.Colour(p.x / r + 0.5, p.y / r + 0.5, p.z / r + 0.5);

        geometry.position(p);
        geometry.colour(c);
    }

    var model = app.gfx.create_model_geo('lines', geometry);

    var n = scene.nodes().add(new two.Node3());
    var it = scene.items().add(new two.Item(n, model, 0, material));
    this.node = n;
}

//var time = app.gfx.time / 2.0; // * 0.001;

var angles = new two.vec3(time * 0.25, time * 0.5, 0.0);
this.node.apply(new two.vec3(0.0), new two.quat(angles));

File: example/xx_three/geom_lines.cpp
module;
#include <infra/Cpp20.h>
#include <xx_three/ex.h>
module two.xxthree;

using namespace two;

EX(xx_geom_lines)
{
#if UI
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	Scene& scene = viewer.m_scene;
#else
	static Scene scene = Scene(app.m_gfx);
	static GfxViewer viewer = GfxViewer(window, scene);
#endif

	constexpr size_t segments = 10000;

	//static Program& program = app.m_gfx.programs().fetch("line");
	static Program& program = app.m_gfx.programs().fetch("solid");

	static Node3* node = nullptr;

	if(init)
	{
		Camera& camera = viewer.m_camera;
		camera.m_fov = 27.f; camera.m_near = 1.f; camera.m_far = 4000.f;
		camera.m_eye.z = 2750.f;
		
		Material& material = app.m_gfx.materials().create("lines", [](Material& m) {
			m.m_program = &program;
			m.m_base.m_shader_color = ShaderColor::Vertex;
		});

		MeshPacker geometry;
		geometry.m_primitive = PrimitiveType::LineStrip;

		float r = 800.f;

		for(size_t i = 0; i < segments; i++)
		{
			vec3 p = vec3(randf(), randf(), randf()) * r - r / 2.f;
			Colour c = to_colour(p / r + 0.5f);

			geometry.position(p);
			geometry.colour(c);
		}

		Model& model = app.m_gfx.create_model_geo("lines", geometry);

		Node3& n = gfx::nodes(scene).add(Node3());
		Item& it = gfx::items(scene).add(Item(n, model, 0U, &material));
		UNUSED(it);
		node = &n;
	}

	const float time = app.m_gfx.m_time / 2.f; // * 0.001;

	vec3 angles = vec3(time * 0.25f, time * 0.5f, 0.f);
	node->apply(vec3(0.f), quat(angles));
}
